File: cs-fundamentals/mutex-semaphore-monitor/mutex_demo.py
import threading
import time

# Shared Resource
shared_counter = 0

# Initialize the Mutex (Lock)
mutex = threading.Lock()

def increment_counter(thread_name):
    global shared_counter

    print(f"[{thread_name}] Waiting for Mutex...")

    # Critical Section: Only one thread can enter this block at a time
    with mutex:
        print(f"[{thread_name}] Acquired Mutex. Processing...")

        # Simulate a time-consuming task to visualize blocking mechanism
        local_copy = shared_counter
        time.sleep(2) # 2 seconds
        shared_counter = local_copy + 1

        print(f"[{thread_name}] Updated counter to {shared_counter}")

    # End of Critical Section (Mutex is automatically released here)

# Initialize threads
threads = []
print("--- Starting Simulation WITH Mutex ---")
for i in range(5):
    t = threading.Thread(target=increment_counter, args=(f"Thread-{i+1}",))
    threads.append(t)
    t.start()

for t in threads:
    t.join()

print(f"Final result: {shared_counter}")

File: cs-fundamentals/mutex-semaphore-monitor/mutex_demo.go
package main

import (
	"fmt"
	"sync"
	"time"
)

// Shared Resource
var sharedCounter int = 0

// Initialize a Mutex to protect the critical section
var mu sync.Mutex

func incrementCounter(id int, wg *sync.WaitGroup) {
	defer wg.Done()

	fmt.Printf("[Goroutine-%d] Waiting for Mutex...\n", id)

	// --- CRITICAL SECTION (Protected) ---
	// Lock the mutex before accessing the shared resource
	// Other goroutines must wait if the mutex is already locked
	mu.Lock()
	
	fmt.Printf("[Goroutine-%d] Acquired Mutex. Processing...\n", id)

	// Safe to access sharedCounter now
	localCopy := sharedCounter
	time.Sleep(2 * time.Second) 
	sharedCounter = localCopy + 1

	fmt.Printf("[Goroutine-%d] Updated counter to %d\n", id, sharedCounter)
	
	// Unlock the mutex to allow other waiting goroutines to enter
	mu.Unlock()
	// ------------------------------------
}

func main() {
	var wg sync.WaitGroup
	fmt.Println("--- Starting Simulation WITH Mutex ---")

	for i := 1; i <= 5; i++ {
		wg.Add(1)
		go incrementCounter(i, &wg)
	}

	wg.Wait()

	fmt.Printf("Final result: %d\n", sharedCounter)
}